src/main/java/jenkins/plugins/slack/SlackUserIdResolver.java

```diff
@@ -0,0 +1,156 @@
+    private static final String SLACK_ID_FIELD = "id";
+
+    private final String authToken;
+    private final CloseableHttpClient httpClient;
+    private final List<MailAddressResolver> mailAddressResolvers;
+
+    public SlackUserIdResolver(String authToken, CloseableHttpClient httpClient, List<MailAddressResolver> mailAddressResolvers) {
+        this.authToken = authToken;
+        this.httpClient = httpClient;
+        this.mailAddressResolvers = (mailAddressResolvers != null) ? mailAddressResolvers : MailAddressResolver.all();
+    }
+
+    public SlackUserIdResolver(String authToken, CloseableHttpClient httpClient) {
+        this(authToken, httpClient, null);
+    }
+
+    public List<String> resolveUserIdsForBuild(AbstractBuild build) {
+        return resolveUserIdsForChangeLogSets(build.getChangeSets());
+    }
+
+    public List<String> resolveUserIdsForRun(Run run) {
+        if (run instanceof RunWithSCM) {
+            RunWithSCM r = (RunWithSCM) run;
+            return resolveUserIdsForChangeLogSets(r.getChangeSets());
+        } else {
+            return Collections.EMPTY_LIST;
+        }
+    }
+
+    public List<String> resolveUserIdsForChangeLogSet(ChangeLogSet changeLogSet) {
+        return Arrays.stream(changeLogSet.getItems())
+                .map(item -> resolveUserId(((Entry) item).getAuthor()))
+                .filter(Objects::nonNull)
+                .collect(Collectors.toList());
+    }
+
+    public List<String> resolveUserIdsForChangeLogSets(List<ChangeLogSet> changeLogSets) {
+        return changeLogSets.stream()
+                .map(changeLogSet -> resolveUserIdsForChangeLogSet(changeLogSet))
+                .flatMap(Collection::stream)
+                .collect(Collectors.toList());
+    }
+
+    public List<String> resolveUserIdsForEmailAddresses(List<String> emailAddresses) {
+        return emailAddresses.stream()
+                .map(address -> resolveUserIdForEmailAddress(address))
+                .filter(Objects::nonNull)
+                .collect(Collectors.toList());
+    }
+
+    public String resolveUserId(User user) {
+        for (MailAddressResolver resolver : mailAddressResolvers) {
+            String emailAddress = resolver.findMailAddressFor(user);
+            if (StringUtils.isNotEmpty(emailAddress)) {
+                String userId = resolveUserIdForEmailAddress(emailAddress);
+                if (StringUtils.isNotEmpty(userId)) {
+                    return userId;
+                }
+            }
+        }
+        return null;
+    }
+
+    public String resolveUserIdForEmailAddress(String emailAddress) {
+        if (StringUtils.isEmpty(emailAddress) || StringUtils.isEmpty(authToken)) {
+            return null;
+        }
+
+        // prepare get method for looking up Slack userId by email address
+        String url = String.format(LOOKUP_BY_EMAIL_METHOD_URL_FORMAT, authToken, emailAddress);
+
+        HttpGet getRequest = new HttpGet(url);
+        try {
```

Comment:
I think this should be:
```suggestion
        try(CloseableHttpResponse response = httpClient.execute(getRequest)) {
```

then you wouldn't need the finally block at the end?